### PR TITLE
Avoid to have to wait to zoom after scrolling

### LINKED ISSUE #17707: High value of WHEEL_ZOOM_DISABLED_TIMEOUT causes delay between scrolling and zooming
Attach (recommended) or Link to PDF file here:
https://people.eecs.berkeley.edu/~krste/papers/riscv-privileged-v1.9.pdf

Configuration:
- Web browser and its version: Mozilla Firefox 122.0.1 (64-bit)
- Operating system and its version: AlmaLinux 8.8
- PDF.js version: builtin into firefox
- Is a browser extension: No

Steps to reproduce the problem:
1. Scroll downwards
2. While scrolling downwards, press ctrl to start zooming

What is the expected behavior?
Immediately zoom into the pdf.

What went wrong?
It takes exactly 2 seconds (measured to a few tenths of a second) before zooming starts. This makes the pdf.js feel unresponsive to user input.

The problem is the WHEEL_ZOOM_DISABLED_TIMEOUT value being set to 1000. For the zoom delay problem it seems to count double for some reason. Settings this value to something like 100 or lower vastly reduces the scroll-to-zoom delay.

### PR DESCRIPTION
**Pull request by calixteman**
Allow to zoom with the wheel once the scrolling is finished.
It's now possible to know that thanks to the scrollend event.

Fixes #17707.

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
> Keep the previous way to work for browsers not supporting the scrollend event.

Do we actually have to care about such browsers, or could we perhaps take the opportunity to simplify the affected code?

Looking at the [MDN compatibility data](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollend_event#browser_compatibility) it seems that *almost* all modern browsers already support this. The only exception is *unsurprisingly* Safari, which has always been the "worst" browser to attempt to support. (Historically it's often been comparatively slow at implement new web-platform features, and over time we've seen a large number issues that only affect Safari.)

**Comment by calixteman** [comment]
> > Keep the previous way to work for browsers not supporting the scrollend event.
> 
> Do we actually have to care about such browsers, or could we perhaps take the opportunity to simplify the affected code?
> 
> Looking at the [MDN compatibility data](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollend_event#browser_compatibility) it seems that _almost_ all modern browsers already support this. The only exception is _unsurprisingly_ Safari, which has always been the "worst" browser to attempt to support. (Historically it's often been quite slow at implementing new web-platform features, and over time we've seen a large number issues that only affect Safari.)

I mostly added this extra code because of:
https://github.com/mozilla/pdf.js/wiki/Frequently-Asked-Questions#legacy-build
and Chrome 92+, when `scrollend` is available in Chrome 114+.
That said if you agree, I'm fully ok to remove this extra stuff.

**Comment by Snuffleupagus** [comment]
> That said if you agree, I'm fully ok to remove this extra stuff.

My reasoning for removing the old code is:
 - We've *never* guaranteed that every feature will be available in older browsers, since some things cannot be polyfilled.
 - These changes don't apply to the PDF.js *library*, but only to the default viewer (which is developed for Firefox).
 - This was intended as nothing more than a work-around for an old Mac-only issue, where the user is pressing the zoom-modifier key *during* scrolling.

**Review by Snuffleupagus — COMMENTED** [review]
Don't we also need to ensure that the new event listeners are removed on "blur" for the `mainContainer`, and also reset `this._isScrolling = false;` then?

**Comment by calixteman** [comment]
> Don't we also need to ensure that the new event listeners are removed on "blur" for the `mainContainer`, and also reset `this._isScrolling = false;` then?

Yes it's possible, but you've to alt+tab during scrolling: it's very unlikely to happen... That said I'll fix that.

**Comment by Snuffleupagus** [comment]
Having tested the preview there's unfortunately cases where the "scrollend" event doesn't seem to fire as expected, which means that mouse-wheel zooming breaks.

 - Case 1:
   1. Click on the scroll-wheel and move the mouse up or down, to trigger scrolling.
   2. Click again on the scroll-wheel, to stop scrolling.
   3. Hold <kbd>Ctrl</kbd> and scroll the mouse-wheel to zoom.
 

 - Case 2:
   1. Hold the scroll-wheel and move the mouse up or down, to trigger scrolling.
   2. Relese the scroll-wheel, to stop scrolling.
   3. Hold <kbd>Ctrl</kbd> and scroll the mouse-wheel to zoom.

**Comment by calixteman** [comment]
It's a Firefox bug:
https://bugzilla.mozilla.org/show_bug.cgi?id=1881974
That said I found a workaround.

**Comment by Snuffleupagus** [comment]
> It's a Firefox bug:
> https://bugzilla.mozilla.org/show_bug.cgi?id=1881974

Nice find, thanks for filing that!

> That said I found a workaround.

Should we perhaps wait a couple of days, to see if there's a quick fix for the upstream bug, before landing a work-around?

File: web/app.js [hunk 1]
```diff
@@ -683,6 +686,43 @@ const PDFViewerApplication = {
...
+        this._lastScrollLeft === mainContainer.scrollLeft
+      ) {
+        return;
+      }
+      mainContainer.removeEventListener("scroll", scroll, {
+        passive: true,
+      });
+      this._isScrolling = true;
+      const scrollend = () => {
+        this._lastScrollTop = mainContainer.scrollTop;
```
Comment (Snuffleupagus) [inline web/app.js:712]: Don't we need to update the left property as well, since scrolling could happen *only horizontally* as well?

**Comment by calixteman** [comment]
> Should we perhaps wait a couple of days, to see if there's a quick fix for the upstream bug, before landing a work-around?

This bug is very likely there for a long time (I tested with a 2020 version) so I'm pretty sure that we won't have a fix soon (except maybe if it's easy enough to fix). So I think we should merge this PR (if everything is ok of course) and we can remove the workaround once it's fixed.

**Review by Snuffleupagus — APPROVED** [review]
r=me, now it seems to work in all cases I've tested; thank you!

**Pull request merged**